### ISSUE TITLE
The core idea of the work

### ISSUE THREAD

**Issue by hellodrx** [zhang943/Face-Adversarial-Attack#1]
Hi, Zhang, 

First, thanks for releasing the code. The code is simple and easy to understand, beautiful work! 

From the code, I see you are trying to build a network to regress some kind of features from different encoders. These features are already generated (correct me if I misunderstood) as the groundtruth. 

I am wondering what's the magic of these features to make you get such low testing-error. I do not have gpus at hand to test the code and since you do not mention the core idea of the work in the project introduction, could you kindly explain it. I think people who see this project would also want to know your idea. 

Thank you for your time. 

Ruoxi

**Comment by zhang943** [comment]
Hi Ruoxi,

I am new to this field. I think our method is similar to the famous algorithm **FGSM**.

The core idea is:

- For each image to be attcked, there is a corresponding mask. A new image can be obtained through `image_a = image + mask`, which is used to fool face recognition system.

- It is well known that the features of the same person's faces should be similar, and the features of different person's face should be different. Since the features the encoders output are vectors, we use cosine dist to measure their similarity. 

- Our goal is to fool face recognition system using `image_a`, so we minimize the distance between `feature_a` and `-feature`. Besides, we add L2 constraint to the mask so that it cannot be too large.

- The face recognition networks(encoders) are pretrained and fixed, while the masks are parameters to be trained. Generating features in advance is to reduce the training time.

Hope it helps you.

Zhe Zhang

**Comment by hellodrx** [comment]
I get it. The features are the outputs of several face-recognition nets (should be like a 512 outputs fc-layer). The hint of the FGSM algorithm also helps me out. 

Thank you very much!
